internal/api/rust/rust.go

```diff
@@ -82,14 +91,93 @@ func NewRustClient(t ct.TestLike, opts api.ClientCreationOpts) (api.Client, erro
 	}
+	if opts.AccessToken != "" { // restore the session
+		session := matrix_sdk_ffi.Session{
+			AccessToken:      opts.AccessToken,
+			UserId:           opts.UserID,
+			DeviceId:         opts.DeviceID,
+			HomeserverUrl:    opts.BaseURL,
+			SlidingSyncProxy: &opts.SlidingSyncURL,
+		}
+		if err := client.RestoreSession(session); err != nil {
+			return nil, fmt.Errorf("RestoreSession: %s", err)
+		}
+		if opts.EnableCrossProcessRefreshLockProcessName == api.ProcessNameNSE {
+			clientSessionDelegate.SaveSessionInKeychain(session)
+			t.Logf("configure NSE client with logged in user: %+v", session)
+			// We purposefully don't SetDelegate as it appears to be unnecessary.
+			notifClientBuilder, err := client.NotificationClient(matrix_sdk_ffi.NotificationProcessSetupMultipleProcesses{})
+			if err != nil {
+				return nil, fmt.Errorf("NotificationClient failed: %s", err)
+			}
+			// weirdly the ffi bindings make new NotificationClientBuilder ffi objects each time you call a chained function,
+			// so we need to destroy() the old one each time or else we can get random rust panics when the finalizer gets
+			// called by Go's GC with "there is no reactor running, must be called from the context of a Tokio 1.x runtime".
+			notifClientBuilder2 := notifClientBuilder.FilterByPushRules()
+			notifClientBuilder.Destroy()
+			c.notifClient = notifClientBuilder2.Finish()
+			notifClientBuilder2.Destroy()
+		}
+	}
+
 	c.Logf(t, "NewRustClient[%s] created client storage=%v", opts.UserID, c.persistentStoragePath)
 	return &api.LoggedClient{Client: c}, nil
 }
 
 func (c *RustClient) Opts() api.ClientCreationOpts {
+	// add access token if we weren't made with it
+	if c.opts.AccessToken == "" && c.FFIClient != nil {
+		session, err := c.FFIClient.Session()
+		if err == nil { // if we ain't logged in, we expect an error
+			c.opts.AccessToken = session.AccessToken
+		}
+	}
 	return c.opts
 }
 
+func (c *RustClient) GetNotification(t ct.TestLike, roomID, eventID string) (*api.Notification, error) {
+	if c.notifClient == nil {
+		t.Errorf("RustClient misconfigured. You can only call GetNotification if this is an NSE process. " +
+			"Ensure opts.EnableCrossProcessRefreshLockProcessName and opts.AccessToken are set!")
+		return nil, fmt.Errorf("misconfigured rust client")
+	}
+	notifItem, err := c.notifClient.GetNotification(roomID, eventID)
+	if err != nil {
+		return nil, fmt.Errorf("GetNotification: %s", err)
+	}
+	// TODO: handle NotificationEventInvite
+	notifEvent := notifItem.Event.(matrix_sdk_ffi.NotificationEventTimeline)
+	// TODO: handle notifications other than messages..
+	evType, err := notifEvent.Event.EventType()
+	if err != nil {
+		return nil, fmt.Errorf("notifItem.Event.EventType => %s", err)
+	}
+	msgLike := evType.(matrix_sdk_ffi.TimelineEventTypeMessageLike)
+	failedToDecrypt := true
+	body := ""
+	switch msg := msgLike.Content.(type) {
+	case matrix_sdk_ffi.MessageLikeEventContentRoomEncrypted:
+		// failedToDecrypt = true
```

Comment:
Because receiving `matrix_sdk_ffi.MessageLikeEventContentRoomEncrypted` _means we failed to decrypt_. Adding the comment is intended to hint that this is the case.